proposals/nnnn-move-function.md

```diff
@@ -0,0 +1,455 @@
+we can use that other binding after we end the lifetime of `x`, as in:
+
+```
+func useX(_ x: SomeClassType) -> () {}
+func consumeX(_ x: __owned SomeClassType) -> () {}
+
+func f() -> () {
+  let x = ...
+  useX(x)
+  let other = x   // other is a new binding used to extend the lifetime of x
+  _ = move(x) // x's lifetime ends
+  useX(other)     // other is used here... no problem.
+  consumeX(other) // other is used here... no problem.
+}
+```
+
+In fact, each variable's lifetime is tracked independently, and gets a separate
+diagnostic if used after move. If we try to compile this:
+
+```
+func useX(_ x: SomeClassType) -> () {}
+func consumeX(_ x: __owned SomeClassType) -> () {}
+
+func f() -> () {
+  let x = ...
+  useX(x)
+  let other = x
+  let _ = move(x)
+  useX(move(other))
+  consumeX(other)
+  useX(x)
+}
+```
+
+we get separate diagnostics for each variable:
+
+```
+test.swift:7:15: error: 'x' used after being moved
+  let x = ...
+          ^
+test.swift:10:11: note: move here
+  let _ = move(x)
+          ^
+test.swift:13:3: note: use here
+  useX(x)
+  ^
+test.swift:9:7: error: 'other' used after being moved
+  let other = x
+      ^
+test.swift:11:8: note: move here
+  useX(move(other))
+       ^
+test.swift:12:3: note: use here
+  consumeX(other)
+  ^
+```
+
+If one applies move to a local `var`, then a new value can be assigned into
+the variable after an old value has been moved out. One can
+begin using the var again after one re-assigns to the var:
+
+```
+func f() {
+  var x = getValue()
+  let _ = move(x)
+  useX(x) // error: no value in x
+  x = getValue()
+  useX(x) // ok, x has a new value here
+}
+```
+
+This follows from move being applied to the binding (`x`), not the value in the
+binding (the value returned from `getValue()`).
+
+We also support applying the move operation to consuming function arguments:
+
+```
+func f(_ x: __owned SomeClassType) {
+    let _ = move(x)
+    useX(x) // !! Error! Use of x after move
+}
+```
+
+Normal arguments are passed by borrow, meaning that the lifetime of the value
+is managed by the caller. Although we could allow `move` on these arguments,
+shortening the syntactic lifetime of the variable, doing so would have no
+practical effect on the value's lifetime at runtime, so we choose to leave this
+disallowed for now.
+
+On the other hand, one can `move` out of an `inout`
+function argument. Like a `var`, the `inout` argument can be reassigned after
+being moved from and used again; however, since the final value of an
+`inout` argument is passed back to the caller, an `inout` argument *must* be
+reassigned by the callee before it returns:
+
+```
+func f(_ buffer: inout Buffer) { // error: 'buffer' not reinitialized after move!
+  let b = move(buffer)           // note: move was here
+  b.deinitialize()
+  ... write code ...
+}                                // note: return without reassigning inout argument `buffer`
```

Comment:
I would put the error on the close brace for purposes of exposition here.  (Even though I understand that's not how the current prototype works.)